packages/host/app/routes/index.ts

```diff
@@ -29,29 +32,103 @@ export default class Index extends Route<void> {
   @service private declare cardService: CardService;
   @service private declare router: RouterService;
   @service private declare operatorModeStateService: OperatorModeStateService;
+  @service declare realm: RealmService;
 
   didMatrixServiceStart = false;
 
-  async beforeModel(transition: Transition): Promise<void> {
+  async fetchCard(url: string) {
+    let resource = getCard(this, () => url);
+    await resource.loaded;
+    return resource.card;
+  }
+
+  async model(params: {
+    cardPath?: string;
+    path: string;
+    operatorModeState: string;
+    operatorModeEnabled: boolean;
+  }): Promise<void> {
+    let { operatorModeState, cardPath } = params;
+
     if (!this.didMatrixServiceStart) {
       await this.matrixService.ready;
       await this.matrixService.start();
       this.didMatrixServiceStart = true;
     }
-    if (!transition.to!.queryParams.operatorModeState) {
+
+    if (!this.matrixService.isLoggedIn) {
+      return; // Show login component
+    }
+
+    let cardUrl: string | undefined;
+
+    if (cardPath) {
+      if (hostsOwnAssets) {
+        cardUrl = new URL(`/${cardPath}`, this.realm.defaultReadableRealm.path)
+          .href;
+      } else {
+        cardUrl = new URL(cardPath, window.location.origin).href;
+      }
+
+      let card = await this.fetchCard(cardUrl); // We are fetching the card here to make sure we have the canonical URL for the card
+
+      if (!card) {
+        // This is a temporary workaround to fix a strange race condition that happens after user tries to see a
+        // card but has to login first. Immediately after login the route will refresh this.fetchCard(cardUrl) will sometimes fail
+        // with a 401 even when the user has permissions to access the card. There has to be a race condition where this happens before
+        // the realm session is established.
+        // TODO: fix the root cause instead of this hack
+        await new Promise((resolve) => setTimeout(resolve, 500)); // Wait for a bit to so that the realm session is established
+        card = await this.fetchCard(cardUrl);
+      }
+
+      let canonicalCardUrl = card?.id;
+
+      if (!canonicalCardUrl) {
+        // TODO: what to do if card is not found, or not accessible?
+        alert(`Card not found: ${cardUrl}`);
```

Comment:
Currently this alerts and then opens the card chooser. Should we design a better notification?

Reply:
totally, we should have an official 404 page. Let's make a ticket to design this